### ISSUE TITLE
Not able to deserialize Enum with default typing after upgrading 2.15.4 -> 2.17.1

### ISSUE THREAD

**Issue by lenrok258** [FasterXML/jackson-databind#4849]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

After bumping up the version to 2.17.1 I'm not longer able to deserialize values serialized with the version 2.15.4.
Moreover, values serialized with the new version (2.17.1) give error when deserializing.

### Version Information

2.17.1

### Reproduction

```java
package com.kze.test;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.jsontype.BasicPolymorphicTypeValidator;
import com.fasterxml.jackson.databind.jsontype.PolymorphicTypeValidator;
import com.fasterxml.jackson.databind.jsontype.impl.StdTypeResolverBuilder;

import java.io.IOException;
import java.util.EnumSet;

import static com.fasterxml.jackson.annotation.JsonTypeInfo.As.PROPERTY;
import static com.fasterxml.jackson.databind.ObjectMapper.DefaultTyping.NON_FINAL;
import static com.kze.test.TestClazz.TestEnum.TEST_ENUM_VALUE;

public class TestClazz {

    public enum TestEnum {
        TEST_ENUM_VALUE
    }

    public static void main(String[] args) throws IOException {

        ObjectMapper mapper = new ObjectMapper();

        final PolymorphicTypeValidator validator = BasicPolymorphicTypeValidator.builder()
                .allowIfSubType("com.kze.test")
                .allowIfSubType("java")
                .build();

        ObjectMapper.DefaultTypeResolverBuilder resolverBuilder
                = new ObjectMapper.DefaultTypeResolverBuilder(NON_FINAL, validator) {
            @Override
            public boolean useForType(JavaType t) {
                return true;
            }
        };

        StdTypeResolverBuilder stdTypeResolverBuilder = resolverBuilder
                .init(JsonTypeInfo.Id.CLASS, null)
                .inclusion(PROPERTY);

        mapper.setDefaultTyping(stdTypeResolverBuilder);
        mapper.configure(JsonParser.Feature.INCLUDE_SOURCE_IN_LOCATION, true);

        EnumSet<TestEnum> input = EnumSet.of(TEST_ENUM_VALUE);
        String inputJson = mapper.writeValueAsString(input);
        System.out.println(inputJson);

        Object inputDeserialized = mapper.readValue(inputJson, Object.class);
        System.out.println(inputDeserialized);
    }

}

``` 

Version 2.15.4 output 
```
["java.util.EnumSet<com.kze.test.TestClazz$TestEnum>",["TEST_ENUM_VALUE"]]
[TEST_ENUM_VALUE]
```

Version 2.17.1 error:
```
["java.util.EnumSet<com.kze.test.TestClazz$TestEnum>",["TEST_ENUM_VALUE"]]
Exception in thread "main" com.fasterxml.jackson.databind.exc.MismatchedInputException: Unexpected token (VALUE_STRING), expected START_ARRAY: need Array value to contain `As.WRAPPER_ARRAY` type information for class com.kze.test.TestClazz$TestEnum
 at [Source: (String)"["java.util.EnumSet<com.kze.test.TestClazz$TestEnum>",["TEST_ENUM_VALUE"]]"; line: 1, column: 56] (through reference chain: java.util.RegularEnumSet[0])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
	at com.fasterxml.jackson.databind.DeserializationContext.wrongTokenException(DeserializationContext.java:1913)
	at com.fasterxml.jackson.databind.DeserializationContext.reportWrongTokenException(DeserializationContext.java:1699)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer._locateTypeId(AsArrayTypeDeserializer.java:141)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer._deserialize(AsArrayTypeDeserializer.java:96)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer.deserializeTypedFromScalar(AsArrayTypeDeserializer.java:66)
	at com.fasterxml.jackson.databind.deser.std.StdScalarDeserializer.deserializeWithType(StdScalarDeserializer.java:66)
	at com.fasterxml.jackson.databind.deser.std.EnumSetDeserializer._deserialize(EnumSetDeserializer.java:281)
	at com.fasterxml.jackson.databind.deser.std.EnumSetDeserializer.deserialize(EnumSetDeserializer.java:246)
	at com.fasterxml.jackson.databind.deser.std.EnumSetDeserializer.deserialize(EnumSetDeserializer.java:24)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer._deserialize(AsArrayTypeDeserializer.java:120)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer.deserializeTypedFromArray(AsArrayTypeDeserializer.java:53)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromAny(AsPropertyTypeDeserializer.java:238)
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializerNR.deserializeWithType(UntypedObjectDeserializerNR.java:112)
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:74)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4905)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3848)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3816)
	at com.kze.test.TestClazz.main(TestClazz.java:52)
```

### Expected behavior

Should be able to deserialized values

**Comment by JooHyukKim** [comment]
Seems similar to the one I am working on ryt now https://github.com/FasterXML/jackson-modules-java8/pull/335

**Comment by lenrok258** [comment]
I'm not sure the route cause is the same. 

In my case serializing Enum does not preceded it with a class name as a first array element
```["TEST_ENUM_VALUE"]```

From some reason older version is able to read it. The newer one throws an error, expecting class name at index 0.

As a proof if you proceed a serialized Enum with a class name at the index "0" it works perfectly fine:
```["com.kze.test.TestClazz$TestEnum", "TEST_ENUM_VALUE"]```

**Comment by JooHyukKim** [comment]
There is `${rootDir}/release-notes` to check what's changed from 2.15.4 -> later version. Could you try out with 2.16.0 as well? It would help track down which version

**Comment by lenrok258** [comment]
Just as a note, using recommended in docs `activateDefaultTyping()` instead of deprecated `setDefaultTyping()` doesn't help.

```java
        ObjectMapper mapper = new ObjectMapper();
        PolymorphicTypeValidator polymorphicTypeValidator =  BasicPolymorphicTypeValidator.builder()
                .allowIfSubType("com.kze.test")
                .allowIfSubType("java")
                .build();
        mapper.activateDefaultTypingAsProperty(polymorphicTypeValidator, EVERYTHING, "type");
        mapper.registerModule(new JavaTimeModule());
```

**Comment by lenrok258** [comment]
> There is `${rootDir}/release-notes` to check what's changed from 2.15.4 -> later version. Could you try out with 2.16.0 as well? It would help track down which version

I've checked 2.17.3 and it behaves exactly the same.

**Comment by cowtowncoder** [comment]
While it probably makes no difference, testing should try the latest patch -- 2.18.2 -- to eliminate possibility of the issue being fixed.

I don't think any change in-between versions would change behavior purposefully, but this is pretty complicated case so probably a side-effect of a fix.

**Comment by lenrok258** [comment]
> While it probably makes no difference, testing should try the latest patch -- 2.18.2 -- to eliminate possibility of the issue being fixed.

I've tested it with the 2.18.2 and it's still not working.

What I found out is the problematic bit is a EnumSet. If I replace it with regular Set it works perfectly fine in all the versions!
In my example attached to this ticket if you change the line 
```java
EnumSet<TestEnum> input = EnumSet.of(TEST_ENUM_VALUE);
``` 
with
```java
Set<TestEnum> input = Set.of(TEST_ENUM_VALUE);
``` 
mapper is able to deserialize json, no excaption is being thrown and the output is:
```
["java.util.ImmutableCollections$Set12",[["com.kze.test.TestClazz$TestEnum","TEST_ENUM_VALUE"]]]
[TEST_ENUM_VALUE]
```

So I guess it is rather rare case as it needs to have EnumSet with enum values serialized and deserialized with default typing enabled.

**Comment by cowtowncoder** [comment]
@lenrok258 good analysis. Yes, not super common problem but definitely something that'd be good to fix.

`EnumSet` is quite special since type id serialized is the only type Jackson adds generic type info on, because otherwise it is impossible to resolve actual type (must use Enum class in constructor).

**Comment by lenrok258** [comment]
@cowtowncoder 
Any ideas for a workaround? 
As values are in a production db I need to be able to deserialize them using newer Jackson.

Reading
```json
["java.util.EnumSet<com.kze.test.TestClazz$TestEnum>",["TEST_ENUM_VALUE"]
```
will always ends up with the Exception.

I've tried to add 
```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NONE)
```
on the EnumSet property but with no luck. Seems like deserializer doesn't care and still expects array wrapper.

**Comment by lenrok258** [comment]
> For urgent case, u may want tp update the varchar "j.u.EnumSet" to Set then modify code to deserialize into Set not EnumSet. But... even better create a clone column and leave the existing data as is.

Unformtunately as the json in db has the `java.util.EnumSet<com.kze.test.TestClazz$TestEnum>` as the type I'm not able to deserialize it into Set.

 I think my only option is to write a migration code which takes all the values from the db and saves them again with EnumSet->Set migrated.

**Comment by lenrok258** [comment]
Just to let you know, applying @JsonTypeInfo(use = NONE) on the enum declaration instead of on EnumSet property seems to be working fine.

```java
@JsonTypeInfo(use = NONE)
public enum TestEnum {
        TEST_ENUM_VALUE
    }
```

Reading 
```json
["java.util.EnumSet<com.kze.test.TestClazz$TestEnum>",["TEST_ENUM_VALUE"]
```
doesn't throw Exception in newer versions of Jackson.

**Comment by JooHyukKim** [comment]
Thank you for sharing workaround @lenrok258! 🙏🏼🙏🏼 
Hopefully I get to find solution tonight (taking a shot at it) 🥲

**Comment by JooHyukKim** [comment]
@lenrok258 Btw, are you able to upgrade to 2.18 version and not 2.17? 2.17 is no longer maintained and only fatal security issues would be back-ported.

**Comment by cowtowncoder** [comment]
Quick note: 2.17 branch is still open not just for security fixes (unlike
1.16). Although each patch will be evaluated for risk so backporting may or
may not be done for any particular one.

**Comment by cowtowncoder** [comment]
Ok, I think this was caused by fix to #4214 which enabled use of (polymorphic) type id on deserialization (see `EnumSetDeserializer`) but NOT on serialization (see `EnumSetSerializer`). It is curious that such a change resolved that particular problem.

Anyway: I think the fix needed is to add handling in `EnumSetSerializer`; I'll see if I can modify #4857 with that idea.
